### ISSUE TITLE
MAINT: Use protected branches to allow auto-merge

### ISSUE THREAD

**Issue by larsoner** [scipy/scipy#16711]
A lot of times I want to "merge when green". I propose we:

1. Set `main` as a protected branch with all CIs required to pass (`Require status checks to pass before merging`)
2. Uncheck the `Include administrators` box in the branch protection settings, which allows admins to override this
3. Turn on the `Allow auto-merge` in the repo general settings

This would allow me to click "Squash and merge when ready" in https://github.com/scipy/scipy/pull/16709 for example. I've used it in a couple of other repos for a couple of weeks and found it pretty useful so far.

My only hesitation is that it seems like there are hours-days where one or two CIs are unreliable or broken. (Ideally CIs would not break like this at all, but given the complexity and broad testing scope of SciPy I guess it's inevitable.) In this case, we would either need to 1) remove these from the branch protection rules for a while so that they can be ignored (annoying), or 2) an administrator would have to merge in these cases (there is a checkbox admins can use in a given PR to say "ignore the requirements and merge anyway").

This would increase burden on admins -- does it seem worth it in terms of the auto-merge tradeoff? It looks like there are 5 people listed as admins who have been active lately (@tylerjereddy, @ev-br, @rgommers, @stefanv, and me) so I guess the burden would mostly fall to us in the "hey admins, there is a broken CI we should ignore, please merge for us" case, so let me know what you think.

These are also just easily toggled checkboxes, so we could try it for a couple of weeks and easily disable it if it's counterproductive.

**Comment by rgommers** [comment]
CI is red about as often as not, for reasons unrelated to the PR. We'd at least have to permanently exclude the pre-release jobs and the lint job. Even then, this is likely to turn into a blocker (we could try and revert if that is true though).

**Comment by rgommers** [comment]
Merge on green is quite nice though, I agree.

**Comment by tupui** [comment]
Protecting branches for direct push and force push is nice, but I would not require a green check as Ralf pointed out.

+1 for the merge when green. I am actually already doing this with SciPy thanks to this browser extension https://github.com/refined-github/refined-github

**Comment by larsoner** [comment]
> +1 for the merge when green

At least to do it with the simple GitHub repo settings, you have to enable "only allow merge when green (but admins can be an exception)" to have the "merge when requirements met" option on PRs. (Or at least for the requirements to be meaningful in the sense we're talking about, i.e., when green)

There are other ways to do it, like allowing an `automerge` label and adding a bot, or browser extensions. But it would be nice to be able to use the GH interface.

Maybe we should wait until they allow adding exceptions / allowing overrides for different privilege levels like `write` instead of just `admin`. Then there wouldn't really be a downside to the GH approach because you can just click a box in the PR saying "yes I really want to ignore the CIs".

**Comment by WarrenWeckesser** [comment]
This looks pretty cool, but please, only use it when you know that the commit message generated by the squash-and-merge will actually be a *good* commit message.  In my experience, that only applies to the simplest pull requests, where there are just one or two commits that already have good messages and there has been no commit churn during the review.  Some of the commit messages that we are getting these days are, simply put, not good.

I'm composing a <strike>curmudgeonly rant about</strike> *plea for improved* commit messages that I'll send to the mailing list in a day or so. A couple links that will be included are NumPy's instructions for ["Writing the commit message"](https://numpy.org/doc/stable/dev/development_workflow.html#writing-the-commit-message) and Step 3 of Python's guide to ["Accepting and Merging a Pull Request"](https://devguide.python.org/getting-started/git-boot-camp/#accepting-and-merging-a-pull-request).  Lately we have strayed very far from these guidelines.

**Comment by larsoner** [comment]
> This looks pretty cool, but please, only use it when you know that the commit message generated by the squash-and-merge will actually be a good commit message.

My experience so far is that the "(squash and) merge when all requirements met" interface is practically the same as the regular interface. That is, you can modify the squash+merge commit message to be whatever you like, just like in the regular squash+merge interface. So I think this concern, while important, actually applies to any PR merging as opposed to being "merge when green"-specific.

> I'm composing a curmudgeonly rant about plea for improved commit messages that I'll send to the mailing list in a day or so.

Sounds good! I am almost certainly guilty of not paying close attention to the squash+merge commit messages...

**Comment by WarrenWeckesser** [comment]
>  That is, you can modify the squash+merge commit message to be whatever you like, just like in the regular squash+merge interface.

Ah, that's great.  Does the "merge when green" setting get disabled if a new commit is added the the pull request?

**Comment by tupui** [comment]
I agree to say that ATM our git history is not perfect. IMHO this is actually due to our use of `merge` instead of `squash merge`. Even if you miss the step of re-writing the message, chances the title of the PR was at least giving some info. While some info might be missing, I find this way more useful than `Merge pull request #... from ...` (yes I know that there is the title of the PR in the next lines, but GUI and such don't show this at first glance as not the first line).

And to the "authorship/personal stats" issue, we can just normalise the use of `Co-authored-by` for reviewers.

**Comment by rgommers** [comment]
> we can just normalise the use of `Co-authored-by` for reviewers.

Please don't do that, review=review and authoring=authoring, they're separate and we typically want a reviewer who is not a (co-)author of the code being merged. There's always some grey zone, if a reviewer starts pushing commits with nontrivial content then they become a co-author; but a few style changes right before merge does not.
